azavea/echo-locator
# Create Django User groups

## Overview

Creates three different `User` groups with variable levels of permissions: Superuser (not actually an explicit group, just `superuser` access), Counselor (admin access, but only able to view/modify `Users` in the `HouseSeeker` group and their respective `UserProfiles`), HouseSeekers (do not have admin access at all). The groups and their permissions are created through a Django management command that's run during `scripts/setup`. Right now `UserProfiles` are empty to be populated in #428, but it seemed easier to setup permissions by creating a basic model to get started so I left it in.

### Notes

- I found programmatically creating groups with assigned permissions through a management command the easiest way for permissions to be setup on app start. It's assumed that `HouseSeeker` should be able to add/delete/edit/view their own `UserProfile`, but since it won't be through the Admin site I left those permissions off for now. If we wanted to add any other permissions (like for the API models) the `MODELS` list in the management command would need to be updated.
- I created a receiver so that when a `User` is added to the `Counselor` group it will automatically update `is_staff` to `True`, and set to `False` when removed from `Counselors`. I'm hoping this makes roles/permissions simpler, but I'm still on the fence if this is worthwhile vs just making sure on `User` creation we assign to group and also set `is_staff` to True. It seemed simpler, so I left it in.
- When a `Counselor` logs in and views a `HouseSeeker` User, all fields will be visible. I originally started taking off the Permissions section since that didn't seem relevant to `Counselor`/something they shouldn't be able to edit, but I began to go down the rabbit hole on ways the Counselor admin should look different than the Superuser and felt like that could end up being it's own task so I put it to the side for now.


## Testing Instructions

 * Run `scripts/setup` so that the new `create_groups` command runs
 * Create a superuser if you haven't already: run `scripts/manage createsuperuser`
 * `scripts/server`
 * Navigate to `/admin` and log in
 * Click on "Groups" and confirm you see `Counselors` and `HouseSeekers`
 * Click on "Users" and create a new `Counselor` user and `HouseSeeker` user:
     * Add a new user and then in profile select "Counselor" in the groups section and save
     * Add a new user and then in profile select "HouseSeeker" in the groups section and save
     * Alternatively can create these users and assign to groups in the Django shell
* Click on the `Counselor` user you just created and confirm the `is_staff` checkbox has been automatically checked in the profile
* Logout and login as the `Counselor` user, confirm:
     * Can login to `/admin/` and see "Users" as well as "User profiles"
     * Only users visible are in `HouseSeeker` group
     * Can not add/delete a `User`
     * Can not add/delete a `UserProfile`
* Logout and try to login to `/admin/ `as a `HouseSeeker` and it should fail
* Run `./scripts/manage test` and confirm all pass


Resolves #427
